crates/ruff_python_formatter/src/comments/format.rs

```diff
@@ -376,3 +359,53 @@ impl Format<PyFormatContext<'_>> for FormatEmptyLines {
         }
     }
 }
+
+/// A helper for normalizing comments efficiently.
+fn normalize_comment<'a>(
+    comment: &'a SourceComment,
+    source: SourceCode<'a>,
+) -> FormatResult<Cow<'a, str>> {
+    let slice = comment.slice();
+    let comment_text = slice.text(source);
+
+    let trimmed = comment_text.trim_end();
+    let trailing_whitespace_len = comment_text.text_len() - trimmed.text_len();
+
+    let Some(content) = trimmed.strip_prefix('#') else {
+        return Err(FormatError::syntax_error(
+            "Didn't find expected comment token `#`",
+        ));
+    };
+
+    if content.is_empty() {
+        return Ok(Cow::Borrowed("#"));
+    }
+
+    // Fast path for correctly formatted comments:
+    // * Start with a `# '.
+    // * Have no trailing whitespace.
+    if trailing_whitespace_len == TextSize::new(0) && content.starts_with(' ') {
+        return Ok(Cow::Borrowed(comment_text));
+    }
+
+    if content.starts_with('\u{A0}') {
+        let trimmed = content.trim_start_matches('\u{A0}');
+
+        // Black adds a space before the non-breaking space if part of a type pragma.
+        if trimmed.trim_start().starts_with("type:") {
+            return Ok(Cow::Owned(std::format!("# \u{A0}{trimmed}")));
```

Comment:
I believe Black's logic is to not normalize the content in that case, meaning we can return `Cow::Borrowed` 
instead


https://github.com/psf/black/blob/b4dca26c7d93f930bbd5a7b552807370b60d4298/src/black/comments.py#L122-L127

Reply:
Hmm. I'll check this out. You might be right (thanks for the link), but I'm wondering if since NBSP isn't in the `COMMENT_EXCEPTIONS` you end up with `'# \u{A0}...` anyway [one the next line](https://github.com/psf/black/blob/b4dca26c7d93f930bbd5a7b552807370b60d4298/src/black/comments.py#L128-L129).

I noticed this last night but it was late, so I could have misunderstood lol https://twitter.com/cnpryer/status/1696349192497541504?s=20

Reply:
I can double check this, but FWIW I returned the comment as-is borrowed:

```diff
diff --git a/crates/ruff_python_formatter/src/comments/format.rs b/crates/ruff_python_formatter/src/comments/format.rs
index 649e80189..756c94ee2 100644
--- a/crates/ruff_python_formatter/src/comments/format.rs
+++ b/crates/ruff_python_formatter/src/comments/format.rs
@@ -452,7 +452,7 @@ fn normalize_comment<'a>(
 
         // Black adds a space before the non-breaking space if part of a type pragma.
         if trimmed.trim_start().starts_with("type:") {
-            return Ok(Cow::Owned(std::format!("# \u{A0}{trimmed}")));
+            return Ok(Cow::Borrowed(comment_text));
         }
```

```
+--- Black␊
++++ Ruff␊
+@@ -8,9 +8,9 @@␊
+ result = 1  # A simple comment␊
+ result = (1,)  # Another one␊
+ ␊
+-result = 1  #  type: ignore␊
++result = 1  # type: ignore␊
+ result = 1  # This comment is talking about type: ignore␊
+-square = Square(4)  #  type: Optional[Square]␊
++square = Square(4)  # type: Optional[Square]␊
+ ␊
+ ␊
+ def function(a: int = 42):␊
```